tephra-core/src/main/java/co/cask/tephra/distributed/ElasticPool.java

```diff
@@ -134,11 +139,9 @@ public T obtain() throws E {
    * @param element the element to be returned
    */
   public void release(T element) {
-    synchronized (this) {
-      this.recycle(element);
-      this.elements.add(element);
-      this.notify();
-    }
+    recycle(element);
+    elements.add(element);
+    semaphore.release();
```

Comment:
It is a good idea to add all releases in a finally block, so that we never lose permits.